### PR TITLE
Provide better diagnostics if a fetch fails

### PR DESCRIPTION
**Pull request by mattcaswell**
Fetch failures are a common problem. If that happens it is useful to have more detailed information about the fetch, such as the algorithm that was being fetched, along with the properties and whether we were using the default libctx or not.

### CODE REVIEW

File: crypto/evp/evp_fetch.c [hunk 1]
```diff
@@ -294,9 +294,25 @@ void *evp_generic_fetch(OPENSSL_CTX *libctx, int operation_id,
...
+        ERR_add_error_data(6,
+                           openssl_ctx_is_default(libctx)
+                           ? "Default library context"
+                           : "Non-default library context",
+                           ", Algorithm (",
+                           name == NULL ? "<null>" : name,
+                           "), Properties (",
+                           properties == NULL ? "<null>" : properties,
+                           ")");
+#endif
```
Comment (levitte) [inline crypto/evp/evp_fetch.c]:
``` C
        int code = EVP_R_FETCH_FAILED;

#ifdef FIPS_MODE
        ERR_raise(ERR_LIB_EVP, code);
#else
        ERR_raise_data(ERR_LIB_EVP, code,
                       "%s, Algorithm (%s), Properties (%s)",
                       (openssl_ctx_is_default(libctx)
                        ? "Default library context"
                        : "Non-default library context"),
                       name = NULL ? "<null>" : name,
                       properties == NULL ? "<null>" : properties);
#endif
```

Dunno about you, but I find that a tad more readable...

File: crypto/evp/evp_fetch.c [hunk 2]
```diff
@@ -314,9 +330,32 @@ void *evp_generic_fetch_by_number(OPENSSL_CTX *libctx, int operation_id,
...
+                                        free_method);
+
+    if (ret == NULL) {
+        EVPerr(0, EVP_R_FETCH_FAILED);
+#ifndef FIPS_MODE
+        {
+            OSSL_NAMEMAP *namemap = ossl_namemap_stored(libctx);
+            const char *name = (namemap == NULL)
+                               ? NULL
+                               : ossl_namemap_num2name(namemap, name_id, 0);
```
Comment (levitte) [inline crypto/evp/evp_fetch.c]: There's a `evp_first_name()`...

File: crypto/evp/exchange.c [hunk 3]
```diff
@@ -229,9 +229,12 @@ int EVP_PKEY_derive_init(EVP_PKEY_CTX *ctx)
 
     /*
      * Because we cleared out old ops, we shouldn't need to worry about
-     * checking if exchange is already there.
+     * checking if exchange is already there. The fetch can fail in the legacy
+     * case.
      */
+    ERR_set_mark();
     exchange = EVP_KEYEXCH_fetch(ctx->libctx, supported_exch, ctx->propquery);
+    ERR_pop_to_mark();
```
Comment (levitte) [inline crypto/evp/exchange.c]: If you look, you'll find `ERR_set_mark()` and `ERR_pop_to_mark()` are already used, but not so tightly around this specific call.

Comment (levitte) [reply crypto/evp/exchange.c]: That goes for all places you added them

File: providers/common/provider_util.c [hunk 4]
```diff
@@ -76,12 +76,17 @@ int ossl_prov_cipher_load_from_params(PROV_CIPHER *pc,
...
     pc->cipher = pc->alloc_cipher = EVP_CIPHER_fetch(ctx, p->data, propquery);
     /* TODO legacy stuff, to be removed */
 #ifndef FIPS_MODE /* Inside the FIPS module, we don't support legacy ciphers */
     if (pc->cipher == NULL)
         pc->cipher = EVP_get_cipherbyname(p->data);
 #endif
+    if (pc->cipher != NULL)
+        ERR_pop_to_mark();
+    else
+        ERR_clear_last_mark();
```
Comment (levitte) [inline providers/common/provider_util.c]: Ok, perhaps not *all* places...

File: ssl/ssl_ciph.c [hunk 5]
```diff
@@ -322,6 +322,7 @@ int ssl_load_ciphers(SSL_CTX *ctx)
     const ssl_cipher_table *t;
 
     disabled_enc_mask = 0;
+
```
Comment (levitte) [inline ssl/ssl_ciph.c]: Spurious?

File: ssl/ssl_lib.c [hunk 6]
```diff
@@ -5862,8 +5864,11 @@ const EVP_CIPHER *ssl_evp_cipher_fetch(OPENSSL_CTX *libctx,
     }
 #endif
 
-    /* Otherwise we do an explicit fetch */
-    return EVP_CIPHER_fetch(libctx, OBJ_nid2sn(nid), properties);
+    /* Otherwise we do an explicit fetch. This may fail and that could be ok */
```
Comment (levitte) [inline ssl/ssl_lib.c]: Uhm, why is that ok?

File: crypto/evp/evp_fetch.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/evp/evp_fetch.c]: Ok - I changed it to that.

File: crypto/evp/evp_fetch.c (hunk 2 above)
Comment (mattcaswell) [reply crypto/evp/evp_fetch.c]: That requires you to have a provider reference - which we don't here.

File: crypto/evp/exchange.c (hunk 3 above)
Comment (mattcaswell) [reply crypto/evp/exchange.c]: Ah - thanks. I missed that. I rolled this change back - and a number of the others.

File: ssl/ssl_ciph.c (hunk 5 above)
Comment (mattcaswell) [reply ssl/ssl_ciph.c]: Removed

File: ssl/ssl_lib.c (hunk 6 above)
Comment (mattcaswell) [reply ssl/ssl_lib.c]: Because in libssl we sometimes try to get a cipher which we know is not available on all platforms.

**Comment by mattcaswell** [comment]
Fixups pushed addressing the comments. Please take a look.

File: crypto/evp/evp_fetch.c (hunk 2 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: Ah, right...

File: ssl/ssl_lib.c (hunk 6 above)
Comment (levitte) [reply ssl/ssl_lib.c]: Ah, all right

File: crypto/evp/exchange.c (hunk 3 above)
Comment (levitte) [reply crypto/evp/exchange.c]: It seems you forgot about this

File: crypto/evp/m_sigver.c [hunk 7]
```diff
@@ -110,10 +110,13 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
...
     /*
      * Because we cleared out old ops, we shouldn't need to worry about
-     * checking if signature is already there.
+     * checking if signature is already there. The fetch can fail in the legacy
+     * case.
      */
+    ERR_set_mark();
     signature = EVP_SIGNATURE_fetch(locpctx->libctx, supported_sig,
                                     locpctx->propquery);
+    ERR_pop_to_mark();
```
Comment (levitte) [inline crypto/evp/m_sigver.c]: ... and this...

File: crypto/evp/signature.c [hunk 8]
```diff
@@ -391,10 +391,12 @@ static int evp_pkey_signature_init(EVP_PKEY_CTX *ctx, int operation)
 
     /*
      * Because we cleared out old ops, we shouldn't need to worry about
-     * checking if signature is already there.
+     * checking if signature is already there. This can fail in the legacy case.
      */
+    ERR_set_mark();
     signature =
         EVP_SIGNATURE_fetch(ctx->libctx, supported_sig, ctx->propquery);
+    ERR_pop_to_mark();
```
Comment (levitte) [inline crypto/evp/signature.c]: ... and here as well...

File: crypto/evp/pmeth_fn.c [hunk 9]
```diff
@@ -71,10 +71,12 @@ static int evp_pkey_asym_cipher_init(EVP_PKEY_CTX *ctx, int operation)
 
     /*
      * Because we cleared out old ops, we shouldn't need to worry about
-     * checking if cipher is already there.
+     * checking if cipher is already there. This can fail in the legacy case.
      */
+    ERR_set_mark();
     cipher =
         EVP_ASYM_CIPHER_fetch(ctx->libctx, supported_ciph, ctx->propquery);
+    ERR_pop_to_mark();
```
Comment (levitte) [inline crypto/evp/pmeth_fn.c]: ... oh, and here too

File: ssl/ssl_lib.c [hunk 10]
```diff
@@ -5913,7 +5920,10 @@ const EVP_MD *ssl_evp_md_fetch(OPENSSL_CTX *libctx,
 #endif
 
     /* Otherwise we do an explicit fetch */
-    return EVP_MD_fetch(libctx, OBJ_nid2sn(nid), properties);
+    ERR_set_mark();
+    md = EVP_MD_fetch(libctx, OBJ_nid2sn(nid), properties);
+    ERR_pop_to_mark();
+    return md;
```
Comment (levitte) [inline ssl/ssl_lib.c]: If `md` is NULL at this point, is that an error or not?  If it is, `ERR_set_mark()` / `ERR_pop_to_mark()` is wrong

**Comment by xkqian** [comment]
Does the travis-ci failure of 33398.5 affect the result? I have one new patch which has the same travis-ci failure with 33398.5. Even my patch targets on the aarch64, I am no idea why it fails on x86-64...

File: providers/common/provider_util.c [hunk 11]
```diff
@@ -76,12 +76,17 @@ int ossl_prov_cipher_load_from_params(PROV_CIPHER *pc,
         return 0;
 
     EVP_CIPHER_free(pc->alloc_cipher);
+    ERR_set_mark();
```
Comment (levitte) [inline providers/common/provider_util.c]: Travis tells that you need to include `<openssl/err.h>` in this file.

**Comment by mattcaswell** [comment]
It helps if I actually commit the changes that I made. Sorry!

@levitte - please look again

File: ssl/ssl_lib.c (hunk 10 above)
Comment (mattcaswell) [reply ssl/ssl_lib.c]: No. It's not an error. Sometimes we attempt to get an MD that isn't available everywhere (e.g. GOST mds). We don't know if it's there until we try and get hold of it. If this function returns NULL we just ignore the digest and move on to the next one.

**Comment by mattcaswell** [comment]
> Does the travis-ci failure of 33398.5 affect the result? I have one new patch which has the same travis-ci failure with 33398.5. Even my patch targets on the aarch64, I am no idea why it fails on x86-64...

No. Its a timeout on that job. Its a recurring problem that we haven't figured out how to fix yet.

**Comment by mattcaswell** [comment]
@levitte - any more comments on this?

File: ssl/ssl_lib.c (hunk 10 above)
Comment (levitte) [reply ssl/ssl_lib.c]: Okie, fair enough

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**